### ISSUE TITLE
Ignore some specific -Wmaybe-uninitialized c_vector warnings

### ISSUE THREAD

**Issue by fcooper8472** [Chaste/Chaste#231]
For years now, every new version of GCC has been even more picky with thinking certain c_vector instances may be uninitialized.

Some of these warnings have been correct, but increasingly large numbers appear to be spurious.
Examples can be found on #205 (related to issue #197).

Following discussion, we have decided that as part of the BBSRC Milestone 2.2, we will investigate migrating from Boost uBLAS to Eigen.
As an interim measure, where appropriate, we will selectively ignore the warnings where they are judged to be spurious.

I intend to achieve this by creating a new header with some macros that will resolve to something like:

```cpp
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"

// Code that triggers the warning

#pragma GCC diagnostic pop
```

The header will ensure that these macros are only defined for GCC, so some empty macros will be created for other compilers.

This is a highly targeted way of ensuring that only very specific lines of code have the diagnostic ignored, and I propose only using this approach judiciously on a case-by-case basis where we have carefully examined the code emitting the warning.

**Comment by ptheywood** [comment]
`-Wmaybe-uninitialized` requires compiler optimisations to be enabled to function in GCC, so the GCC 12 build in the ubuntu-compilers workflow doesn't trigger these warnings.

https://github.com/Chaste/Chaste/blob/b98110e9d82a10460375695d02b4040047da1734/.github/workflows/ubuntu-compilers.yml#L33-L35

A local GCC 11.4.0 build did not trigger any `-Wmaybe-uninitialized` warnings, while 12.3.0 does (84 occurences, 46 unique)

12.3.0 also emits a `-Wnonnull` warning three times for `chaste/cell_based/src/cell_based_pde/pdes/AveragedSourceEllipticPde.cpp:117` which you've already raised in #205

**Comment by fcooper8472** [comment]
See also #280

**Comment by fcooper8472** [comment]
When adding the macros, it became apparent that they would be required in a large number (somewhere around 100) places.

The decision was taken to, instead, globally turn off the warnings for specific GCC versions, and only when compiling in Release mode. This was implemented here: b94d590b837324c3cd4c0e5ec98beb123a96c9f4

The possibility that this situation is a genuine defect is low, due to:

1. other compilers not issuing any similar warnings, and GCC not warning in Debug -> high likelihood there is no fundamental issue with boost ublas
2. all unit tests pass, including on GCC in Release mode -> high likelihood we are not systematically reading from uninitialized memory

In addition, extra tests were added to `TestNode` that copy-assign to `c_vectors` from the result of `rGetLocation()`, carefully checking in multiple dimensions that memory is not being clobbered by successive assignments: d18f283ac0c73e7e5bfdd5f701a2c2b90f2f7944

**Comment by jmpf** [comment]
> 
> In addition, extra tests were added to `TestNode` that copy-assign to `c_vectors` from the result of `rGetLocation()`, carefully checking in multiple dimensions that memory is not being clobbered by successive assignments: [d18f283](https://github.com/Chaste/Chaste/commit/d18f283ac0c73e7e5bfdd5f701a2c2b90f2f7944)

Cool
